### DISCUSSION TITLE
How can I train a Custom NER model on Jupyter Notebook without using CLI?

### DISCUSSION THREAD

**Discussion by vyshv** [explosion/spaCy#7576] [Help: Coding & Implementations]
Hi everyone! I understand from reading the spacy documentation that under V3.0, it is recommended that the training happens on the Command Line Interface. However, is it possible that I continue doing my training on my notebook, internally on Python, the way it was done before? My train data is in the same format that spacy v2 would take.

As of now, I have attempted training it as such, but there is no improvement in the results pre-training and post-training. If anything, the values are actually the exact same. 

My code looks like this:

```
nlp = spacy.load("en_core_web_trf")
entity_ruler = nlp.add_pipe('entity_ruler', before = 'ner') # entity ruler for recognising other entities with regex
entity_ruler.add_patterns(patterns) # where patterns are the different regexes for the entities

ner = nlp.get_pipe("ner")

for text, annotations in data: # adding labels
    for ent in annotations.get("entities"):
        ner.add_label(ent[2])

# get names of other pipes to disable them during training
pipe_exceptions = ["ner", "transformers"] # I understand that under v3.0, the ner component follows transformer so I kept it
other_pipes = [pipe for pipe in nlp.pipe_names if pipe not in pipe_exceptions]

with nlp.disable_pipes(*other_pipes):
    optimizer = nlp.resume_training()
    for itn in range(20):
        random.shuffle(data)
        losses = {}
        for batch in minibatch(data, size=compounding(4.0, 16.0, 1.001)):
            for text, annotations in batch:
                doc = nlp.make_doc(text)
                example = Example.from_dict(doc, annotations)
                nlp.update([example], sgd = optimizer, drop=0.4, losses=losses)
            print ('losses:', losses)

```

Can anyone let me know if this method is correct? Or how I can improve upon this? 

Thank you!

**Comment by polm** [comment]
It looks like you're disabling the Transformer layer. 

    pipe_exceptions = ["ner", "transformers"]

Note that the layer is called "transformer", not "transformers". Because of the way you generate your list of other pipes the Transformer layer will be disabled. 

Is that your issue?

**Reply by polm** [reply]
As a side note, you can still use the command line interface in a Jupyter notebook, you just issue shell commands starting with `!`.
